Adapt scheduler to use epoll cycle callbacks

mihails-strasuns opened the pull request:
@matthias-wende-sociomantic I was not able to finish this to the state that passes tests today but this should explain what I am trying to do. Lot of stuff needs to be changed and deprecation process is pain, but I do like how it is fleshing out design-wise.

Commit:
Tweak DynamicQueue tests

To check intended use case better

matthias-wende-sociomantic commented on src/ocean/io/select/EpollSelectDispatcher.d:
```diff
@@ -802,6 +803,58 @@ public class EpollSelectDispatcher : IEpollSelectDispatcherInfo
+            // next cycle"
+
+            for (int i; i < this.select_cycle_callbacks.length(); ++i)
+            {
```
When the next callback might add itself to the queue this loop is infinite, right?
Assigning `this.select_cycle_callbacks.length()` to a variable and compare against this would solve this problem.

matthias-wende-sociomantic replied:
OK, I was to slow :). Just realized that I didn't send my review yesterday and now it's obsolete :).

mihails-strasuns replied:
You caught the mistake, that is what matters ;)

gavin-norman-sociomantic reviewed (commented):
Aside from the slack discussion re: deprecating `eventLoop`, the basic concept LGTM.

Commit:
Enhance toContextDg

Commit:
Switch scheduler to use `onCycleEnd`

Deprecates all symbols that refer to the old design.

matthias-wende-sociomantic commented on src/ocean/io/select/EpollSelectDispatcher.d:
```diff
@@ -737,16 +769,20 @@ public class EpollSelectDispatcher : IEpollSelectDispatcherInfo
-
         if (select_cycle_hook !is null)
-            caller_work_pending = select_cycle_hook();
+            this.onCycleEnd(select_cycle_hook);
```
This should run more then once :).

mihails-strasuns commented:
Tweaked design based on discussion with @matthias-wende-sociomantic in Slack. Should be review-ready no, apart from missing release notes.

matthias-wende-sociomantic reviewed (commented):
Design wise I'm happy with the changes :). I'm a bit confused about the order of the commits though and by that I mean introducing the `select_cycle_callbacks` with bool return value first adapting everything to work with it and then changing the semantics to `void` callbacks at the end.

I'd rather prefer to introduce the finalized `select_cycle_callbacks` first and then use it for changing the `delayed_resume`.

One suggestion to get a better structure could be:
- Adapt my existing PR (#664) to introduce the new `select_cycle_callbacks`
- Use this PR for `delayed_resume` and change the `select_cycle_hook` to become part of the callbacks
(these 2 would be exactly the state of this PR)
- Use the `select_cycle_callbacks` for throttler resume termination hook.
(this is the reason why this change was proposed at the beginning)

Apart from structuring I haven't looked deep into code details yet. I'll do it later (i.e. when I have more time).

mihails-strasuns commented:
Order of commits is definitely not final - I left it this way for review because it follows thought flow and I felt that it is easier to review this way.

You can pick relevant bits into #664 (adding release notes and stuff 👼) and I will rebase after it is merged.

mihails-strasuns commented:
Rebased, added release notes, squashed related commits. Also tweaked implementation a bit to avoid duplicate cycle callbacks in scheduler.

After this, the only important remaining bit is to change throttledtaskpool to use cycle callbacks too - will do that in a separate PR.

gavin-norman-sociomantic commented on src/ocean/core/TypeConvert.d:
```diff
@@ -328,9 +329,9 @@ unittest
 *******************************************************************************/
 
-void delegate() toContextDg ( alias F ) ( void* context )
+ReturnTypeOf!(F) delegate() toContextDg ( alias F ) ( void* context )
```
The Returns doc should be updated.

gavin-norman-sociomantic commented on src/ocean/task/Scheduler.d:
```diff
@@ -113,6 +114,15 @@ final class Scheduler : IScheduler
+
+    ***************************************************************************/
+
+    size_t cycle_pending_task_count;
```
Should this be private?

gavin-norman-sociomantic commented on src/ocean/task/Scheduler.d:
```diff
@@ -678,32 +673,10 @@ final class Scheduler : IScheduler
-
-        Helper method which combines recurring pattern of resuming some task
-        and handling potential exceptions.
+        if (queued || this.fiber_pool.limit() == this.fiber_pool.num_busy())
```
I don't get the use of `queued` here. Could you add a comment explaining?

mihails-strasuns replied:
It was wrong artifact of old design. Updated to match current approach.

gavin-norman-sociomantic commented on src/ocean/task/Scheduler.d:
```diff
@@ -678,32 +673,10 @@ final class Scheduler : IScheduler
-        Helper method which combines recurring pattern of resuming some task
-        and handling potential exceptions.
+        if (queued || this.fiber_pool.limit() == this.fiber_pool.num_busy())
+            this.epoll.onCycleEnd(&this.selectCycleHook);
```
Should ` this.select_cycle_hook_registered` not be set to true, in this case? (The following line sets it to false.)

mihails-strasuns replied:
Was missing `else` ;)

gavin-norman-sociomantic replied:
Sneaky ;)

gavin-norman-sociomantic reviewed (approved):
With the last tweaks, LGTM. I'll leave it for @matthias-wende-sociomantic to give the final approval, though, as he has a deeper knowledge of the current issues.

matthias-wende-sociomantic reviewed (approved):
I hereby give my final approval! :grin:

gavin-norman-sociomantic commented:
Excellent work chaps!
